multiapps-controller-process/src/test/java/org/cloudfoundry/multiapps/controller/process/steps/CreateOrUpdateStepWithExistingAppTest.java

```diff
@@ -1,388 +1,305 @@
-                                                    .build();
-        }
-
+    static Stream<Arguments> testHandleApplicationEnv() {
+        return Stream.of(Arguments.of(Map.of("foo", "bar"), Map.of("foo1", "bar2"), true, true),
+                         Arguments.of(Map.of("foo", "bar"), Map.of("foo1", "bar2"), false, true),
+                         Arguments.of(Map.of("foo", "bar"), Map.of("foo", "bar"), true, false),
+                         Arguments.of(Map.of("foo", "bar"), Map.of("foo", "bar"), false, false));
     }
 
-    private static class SimpleService {
-        final String name;
+    @ParameterizedTest
+    @MethodSource
+    void testHandleApplicationEnv(Map<String, String> existingAppEnv, Map<String, String> newAppEnv, boolean keepExistingEnv,
+                                  boolean expectedUserPropertiesChanged) {
+        CloudApplication existingApplication = getApplicationBuilder(keepExistingEnv).env(existingAppEnv)
+                                                                                     .build();
+        CloudApplicationExtended application = getApplicationBuilder(keepExistingEnv).env(newAppEnv)
+                                                                                     .build();
+        prepareContext(application, false);
+        prepareClient(existingApplication);
+
+        step.shouldPrettyPrint = () -> false;
+        step.execute(execution);
 
-        SimpleService(String name) {
-            this.name = name;
+        assertStepFinishedSuccessfully();
+        assertEquals(expectedUserPropertiesChanged, context.getVariable(Variables.USER_PROPERTIES_CHANGED));
+        if (expectedUserPropertiesChanged) {
+            Map<String, String> expectedEnvMap = buildExpectedEnvMap(existingAppEnv, newAppEnv, keepExistingEnv);
+            verify(client).updateApplicationEnv(eq(APP_NAME), eq(expectedEnvMap));
+            return;
         }
+        verify(client, never()).updateApplicationEnv(eq(APP_NAME), anyMap());
+    }
 
-        CloudServiceInstanceExtended toCloudService() {
-            return ImmutableCloudServiceInstanceExtended.builder()
-                                                        .metadata(ImmutableCloudMetadata.builder()
-                                                                                        .guid(NameUtil.getUUID(name))
-                                                                                        .build())
-                                                        .name(name)
-                                                        .build();
+    private Map<String, String> buildExpectedEnvMap(Map<String, String> existingAppEnv, Map<String, String> newAppEnv,
+                                                    boolean keepExistingEnv) {
+        if (!keepExistingEnv) {
+            return newAppEnv;
         }
+        return Stream.concat(existingAppEnv.entrySet()
```

Comment:
Wouldn't it be simpler to do `existingAppEnv.putAll(newAppEnv)` (or vice versa) and return the map?
Or just put them in a new map